File: isma-compiler/hsm-core/src/main/java/ru/nstu/isma/core/hsm/service/PDEInitialValueCalculator.java
package ru.nstu.isma.core.hsm.service;

import ru.nstu.isma.core.hsm.exp.*;
import ru.nstu.isma.core.hsm.var.HMConst;
import ru.nstu.isma.core.hsm.var.HMUnnamedConst;
import ru.nstu.isma.core.hsm.var.HMVariable;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

/**
 * by
 * Bessonov Alex.
 * Date: 14.11.13 Time: 0:21
 */
public class PDEInitialValueCalculator {

    private static final HashSet<String> availableFunctions = new HashSet<>();

    static {
        availableFunctions.add("cos");
        availableFunctions.add("sin");
        availableFunctions.add("pow");
        availableFunctions.add("exp");
        availableFunctions.add("sqrt");
    }

    public static Double calculate(HMConst c) {
        if (isCalculated(c) || c instanceof HMUnnamedConst) {
            return c.getValue();
        }
        HMExpression constExp = c.getRightPart();
        validateExpression(constExp);
        avoidLoopCheck(constExp, null);
        Double value = calcExpression(constExp);
        c.setValue(value);
        return value;
    }

    private static void validateExpression(HMExpression exp) {
        List<EXPToken> tokens = exp.getTokens();
        // проверяем что все токены - константы или функции
        // проверяем что калькулятор "знает" применяемые функции
        for (EXPToken t : tokens) {
            if (t instanceof EXPFunctionOperand f) {
                if (!availableFunctions.contains(f.getName())) {
                    // TODO семантика
                    throw new RuntimeException("Calculator don't undastand function " + f.getName());
                }
                for (HMExpression fExp : f.getArgs()) {
                    validateExpression(fExp);
                }
            } else if (t instanceof EXPOperand) {
                HMVariable var = ((EXPOperand) t).getVariable();
                if (!(var instanceof HMConst)) {
                    // TODO обработка ошибки (семантика)
                    throw new RuntimeException("Calculator don't undastand variable " + var.getCode());
                }
            }
        }
    }

    private static void avoidLoopCheck(HMExpression expression, HashSet<String> prevCheck) {

        var current = new HashSet<String>();
        if (prevCheck != null) {
            current.addAll(prevCheck);
        }
        List<EXPToken> tokens = expression.getTokens();
        for (EXPToken token : tokens) {
            if (token instanceof EXPFunctionOperand f) {
                for(var fExp : f.getArgs()) {
                    avoidLoopCheck(fExp, current);
                }
            } else if (token instanceof EXPOperand) {
                HMConst var = (HMConst) ((EXPOperand) token).getVariable();
                if (current.contains(var.getCode())) {
                    // TODO получили петлю -  семантика обработка ошибки
                    throw new RuntimeException("Const contain loop!");
                } else {
                    if (!(var instanceof HMUnnamedConst) && var.getValue()==null) {
                        avoidLoopCheck(var.getRightPart(), current);
                    }
                }
            }
        }
    }

    private static Double calcExpression(HMExpression expression) {
        var tokens = expression.getTokens();

        var stack = new Stack<Double>();
        for (EXPToken token : tokens) {
            if (token instanceof EXPFunctionOperand f) {
                stack.push(calculateFunction(f));
            } else if (token instanceof EXPOperand o) {
                var c = (HMConst) o.getVariable();
                stack.push(calculate(c));
            } else if (token instanceof EXPOperator) {
                doOperation(stack, (EXPOperator) token);
            }
        }
        return stack.pop();
    }

    private static Double calculateFunction(EXPFunctionOperand function) {
        var args = new LinkedList<Double>();
        for (HMExpression exp : function.getArgs()) {
            args.add(calcExpression(exp));
        }
        return switch (function.getName()) {
            case "cos" -> Math.cos(Math.toRadians(args.get(0)));
            case "sin" -> Math.sin(Math.toRadians(args.get(0)));
            case "exp" -> Math.exp(args.get(0));
            case "pow" -> Math.pow(args.get(0), args.get(1));
            case "sqrt" -> Math.sqrt(args.get(0));
            default -> throw new RuntimeException("I undastand function " + function.getName());
        };
    }

    private static void doOperation(Stack<Double> values, EXPOperator o) {
        double result = 0;
        if (o.getType() != EXPOperator.Type.ALGEBRAIC) {
            // TODO семантическая ошибка
        }
        if (o.getArity() == EXPOperator.ArityType.BINARY) {
            double v1 = values.pop();
            double v2 = values.pop();
            EXPOperator.Code c = o.getCode();
            switch (c) {
                case ADDITION:
                    result = v1 + v2;
                    break;
                case DIVISION:
                    result = v2 / v1;
                    break;
                case SUBTRACTION:
                    result = v2 - v1;
                    break;
                case MULTIPLICATION:
                    result = v1 * v2;
                    break;
                default:
                    // TODO неизвестный операнд
                    break;
            }

        } else if (o.getArity() == EXPOperator.ArityType.UNARY) {
            double v1 = values.pop();
            EXPOperator.Code c = o.getCode();
            switch (c) {
                case ADDITION:
                    result = +v1;
                case SUBTRACTION:
                    result = -v1;
                    break;
            }
        }
        values.push(result);
    }

    private static boolean isCalculated(HMConst c) {
        return c.getValue() != null;
    }

}

File: isma-next-hsm/src/main/kotlin/ru/nstu/isma/hsm/service/PDEInitialValueCalculator.kt
package ru.nstu.isma.hsm.service

import ru.nstu.isma.hsm.`var`.HMConst
import ru.nstu.isma.hsm.`var`.HMUnnamedConst
import ru.nstu.isma.hsm.exp.EXPFunctionOperand
import ru.nstu.isma.hsm.exp.EXPOperand
import ru.nstu.isma.hsm.exp.EXPOperator
import ru.nstu.isma.hsm.exp.HMExpression
import java.util.*
import kotlin.math.*

/**
 * by
 * Bessonov Alex.
 * Date: 14.11.13 Time: 0:21
 */
object PDEInitialValueCalculator {
    private val availableFunctions = HashSet<String?>()
    fun calculate(c: HMConst?): Double? {
        if (isCalculated(c) || c is HMUnnamedConst) {
            return c!!.value
        }
        val constExp = c!!.rightPart
        validateExpression(constExp)
        avoidLoopCheck(constExp, null)
        val value = calcExpression(constExp)!!
        c.setValue(value)
        return value
    }

    private fun validateExpression(exp: HMExpression?) {
        val tokens = exp!!.tokens
        // проверяем что все токены - константы или функции
        // проверяем что калькулятор "знает" применяемые функции
        for (t in tokens) {
            if (t is EXPFunctionOperand) {
                val f = t
                if (!availableFunctions.contains(f.name)) {
                    // TODO семантика
                    throw RuntimeException("Calculator don't undastand function " + f.name)
                }
                for (fExp in f.args) {
                    validateExpression(fExp)
                }
            } else if (t is EXPOperand) {
                val `var` = t.variable
                if (`var` !is HMConst) {
                    // TODO обработка ошибки (семантика)
                    throw RuntimeException("Calculator don't undastand variable " + `var`?.code)
                }
            }
        }
    }

    private fun avoidLoopCheck(expression: HMExpression?, prevCheck: HashSet<String?>?) {
        val current = HashSet<String?>()
        if (prevCheck != null) {
            for (s in prevCheck) {
                current.add(s)
            }
        }
        val tokens = expression!!.tokens
        for (token in tokens) {
            if (token is EXPFunctionOperand) {
                for (fExp in token.args) {
                    avoidLoopCheck(fExp, current)
                }
            } else if (token is EXPOperand) {
                val `var` = token.variable as HMConst
                if (current.contains(`var`.code)) {
                    // TODO получили петлю -  семантика обработка ошибки
                    throw RuntimeException("Const contain loop!")
                } else {
                    if (`var` !is HMUnnamedConst && `var`.value == null) {
                        avoidLoopCheck(`var`.rightPart, current)
                    }
                }
            }
        }
    }

    private fun calcExpression(expression: HMExpression?): Double? {
        val tokens = expression!!.tokens
        val stack = Stack<Double>()
        for (token in tokens) {
            when (token) {
                is EXPFunctionOperand -> {
                    stack.push(calculateFunction(token))
                }
                is EXPOperand -> {
                    val c = token.variable as HMConst
                    stack.push(calculate(c))
                }
                is EXPOperator -> {
                    doOperation(stack, token)
                }
            }
        }
        return stack.pop()
    }

    private fun calculateFunction(function: EXPFunctionOperand): Double {
        val args: MutableList<Double?> = LinkedList()
        for (exp in function.args) {
            args.add(calcExpression(exp))
        }
        return when (function.name) {
            "cos" -> {
                cos(Math.toRadians(args[0]!!))
            }
            "sin" -> {
                sin(Math.toRadians(args[0]!!))
            }
            "exp" -> {
                exp(args[0]!!)
            }
            "pow" -> {
                (args[0]!!).pow(args[1]!!)
            }
            "sqrt" -> {
                sqrt(args[0]!!)
            }
            else -> {
                throw RuntimeException("I undastand function " + function.name)
            }
        }
    }

    private fun doOperation(values: Stack<Double>, o: EXPOperator) {
        var result = 0.0
        if (o.type != EXPOperator.Type.ALGEBRAIC) {
            // TODO семантическая ошибка
        }
        if (o.arity == EXPOperator.ArityType.BINARY) {
            val v1 = values.pop()
            val v2 = values.pop()
            val c = o.code
            when (c) {
                EXPOperator.Code.ADDITION -> result = v1 + v2
                EXPOperator.Code.DIVISION -> result = v2 / v1
                EXPOperator.Code.SUBTRACTION -> result = v2 - v1
                EXPOperator.Code.MULTIPLICATION -> result = v1 * v2
                else -> {
                }
            }
        } else if (o.arity == EXPOperator.ArityType.UNARY) {
            val v1 = values.pop()
            val c = o.code
            when (c) {
                EXPOperator.Code.ADDITION -> {
                    result = +v1
                }
                EXPOperator.Code.SUBTRACTION -> {
                    result = -v1
                }
                else -> {}
            }
        }
        values.push(result)
    }

    private fun isCalculated(c: HMConst?): Boolean {
        return c?.value != null
    }

    init {
        availableFunctions.add("cos")
        availableFunctions.add("sin")
        availableFunctions.add("pow")
        availableFunctions.add("exp")
        availableFunctions.add("sqrt")
    }
}